azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java

```diff
@@ -1366,6 +1371,24 @@ public FlowRunnerEventListener() {
       metaData.put("executionId", String.valueOf(flow.getExecutionId()));
       metaData.put("startTime", String.valueOf(flow.getStartTime()));
       metaData.put("submitTime", String.valueOf(flow.getSubmitTime()));
+
+      // Propagate flow properties to Event Reporter
+      if (FlowLoaderUtils.isAzkabanFlowVersion20(flow.getAzkabanFlowVersion())) {
+        // In Flow 2.0, flow has designated properties (defined at its own level in Yaml)
+        propagateMetadataFromProps(metaData, flow.getInputProps(), "flow", flow.getId(), logger);
+      } else {
+        // In Flow 1.0, flow properties are combination of shared properties in individual files (order not defined,
+        // .. because it's loaded by fs list order and put in a HashMap).
+        Props combinedProps = null;
+        for (Props sharedProp : flowRunner.sharedProps.values()) {
```

Comment:
It looks like Props is meant to be hierarchical, with a parent Props, and the current props as a Map. sharedProps is a Map with values as Props, but there doesn't seem to be any order to these, hierarchical or otherwise. Why is a hierarchical graph of Props being created? Would creating the props at the same level make sense?

Would the different Props have the duplicate properties? If so, what is the precedence, if any?

Reply:
In Flow 1.0, flow specifically does not has dedicated properties (except shared props that it maintains for sharing with all jobs). 
The shared props itself is generated by fs ls on property files and puts in HashMap thereby losing any ordering guarantees. 
So, I have simply 'attached' the share props together in one shared props for simplicity (marking one as parent of other). So that when we do a lookup, it looks in one shared prop and if it does not finds the property, then it goes on to look into another (parent). 
Alternatively, I could have iterated over all the shared props one by one .. and stopped when I would have found a property key. But that is still affected by no natural ordering of HashMap key value pairs. 

So, to answer your question, there is no presumed order or precedence within shared props.

Reply:
Could the properties be created at one level then? Right now it is creating a hierarchical order, which indicates an ordering that we're not really guaranteeing. There is the following constructor:

  /**
   * Create properties from maps of properties
   */
  public Props(final Props parent, final Map<String, String>... props) {
    this(parent);
    for (int i = props.length - 1; i >= 0; i--) {
      this.putAll(props[i]);
    }
  }

Reply:
As discussed offline, changing to iterative combination of map to make code reflective of intent (though behavior remains unchanged). 
While at it, figured out that props.putAll takes care of resolving parent configs. so didn't have to write recursive resolution.